spacetime/tests/test_linear_scalar.py

```diff
@@ -127,10 +130,111 @@ def _march():
 
         svr2 = self._build_solver(self.resolution)[-1]
 
-        for it in range(self.nstep*self.cycle):
+        for it in range(self.nstep * self.cycle):
             _march()
             svr2.march_alpha2(steps=1)
-            self.assertEqual(self.svr.get_so0(0).ndarray.tolist(),
-                             svr2.get_so0(0).ndarray.tolist())
+            self.assertEqual(
+                self.svr.get_so0(0).ndarray.tolist(), svr2.get_so0(0).ndarray.tolist()
+            )
+
+
+class NormTC(unittest.TestCase):
+    """
+    Compare linear solver's solution wtih exact solution, because solutions of computation field are not a scalar,
+    therefore compared the norm of difference bewteen solver's solution and exact solution.
+
+    By comparing norm of difference bewteen solver's solution and exact solution we can check if the linear solver is work properly
+    or check if the solver's mathematical model is correct.
+    """
+
+    @staticmethod
+    def _build_solver(resolution):
+        """Create linear solver object for norm test
+        Args:
+            resolution:
+              grid number
+        Return:
+          A well set linear solver object
+        """
+        grid = libst.Grid(0, 4 * 2 * np.pi, resolution)
+        cfl = 1
+        dx = (grid.xmax - grid.xmin) / grid.ncelm
+        dt = dx * cfl
+        svr = libst.LinearScalarSolver(grid=grid, time_increment=dt)
+
+        # Initialize
+        for e in svr.selms(odd_plane=False):
+            if e.xctr < 2 * np.pi or e.xctr > 4 * np.pi:
+                v = 0
+                dv = 0
+            else:
+                v = np.sin(e.xctr)
+                dv = np.cos(e.xctr)
+            e.set_so0(0, v)
+            e.set_so1(0, dv)
+
+        svr.setup_march()
+
+        return svr
+
+    @staticmethod
+    def exact_solution(svr, iter_num, gfun):
+        """Calculate exact solution of the 1D advection equation
+        Args:
+            svr:
+              linear wave solver
+            iter_num:
+              how many step will be executed
+            gfun:
+              system function
+        Return:
+            A ndarray which content current time step solution
+        """
+        v = np.array([])
+        for e in svr.selms(odd_plane=False):
+            x = (
+                e.xctr - iter_num * svr.dt
+            ) % svr.grid.xmax  # dealt with boundary treatment
+            if x < 2 * np.pi or x > 4 * np.pi:
+                v = np.append(v, 0)
```

Comment:
This loop is super inefficient.  Every iteration you are creating a new array and copying over the old one.

Please just use a list and convert it to an array after exiting the loop.